// @web/dev-server-storybook/dist/shared/html/createManagerHtml.js
import fs from 'fs';
import { createBrowserImport } from '../utils.js';
function createManagerImport(rootDir, managerJsPath) {
    if (!fs.existsSync(managerJsPath)) {
        return '';
    }
    const managerImport = createBrowserImport(rootDir, managerJsPath);
    return `import '${managerImport}';`;
}
export function createManagerHtml(storybookConfig, rootDir) {
    var _a;
    const managerImport = createManagerImport(rootDir, storybookConfig.managerJsPath);
    const addonImports = storybookConfig.mainJs.addons
        ? storybookConfig.mainJs.addons.map(a => `import '${a}';`).join('')
        : '';
    return `<!DOCTYPE html>
<html>
  <head>
    <meta charset="UTF-8">
    <meta name="viewport" content="width=device-width, initial-scale=1" />
    <title>Storybook</title>

    <style>
      html,
      body {
        overflow: hidden;
        height: 100%;
        width: 100%;
        margin: 0;
        padding: 0;
      }
    </style>

    <style>
      #root[hidden],
      #docs-root[hidden] {
        display: none !important;
      }
    </style>
    ${(_a = storybookConfig.managerHead) !== null && _a !== void 0 ? _a : ''}
  </head>

  <body>
    <div id="root"></div>
    <div id="docs-root"></div>
    <script type="module">
      import '@web/storybook-prebuilt/manager.js';
      ${managerImport}
      ${addonImports}
    </script>
  </body>
</html>`;
}
